### ISSUE TITLE
ScaNN and BruteForce results do not match on DCN and Deep Layers Architecture

### ISSUE THREAD

**Issue by saihtaungkham** [tensorflow/recommenders#208]
I am trying to apply all the available features on MovieLens dataset, coming up with mixed DCN and Deep layers architecture.
I am not sure whether I have done it in the correct ways. After extracting the user model and passing it to ScaNN layer, somehow the result from BruteForce and ScaNN layers do not match. I hope someone could help me figuring it out what I have missed. Thank you.

<details>
<summary>Model Class</summary>
<pre><code>
class MovielensModel(tfrs.models.Model):
    def __init__(self, rating_weight: float, retrieval_weight: float, use_cross_layer: bool, deep_layer_sizes: List, projection_dim=None) -> None:
        super().__init__()
        
        self.embedding_dimension = 32
        
        str_features = ["movie_id", "user_id", "user_zip_code", "user_occupation_text"]
        int_features = ["user_gender", "bucketized_user_age"]
        
        self._all_features = str_features + int_features
        self._embeddings = {}

        for feature_name in str_features:
            vocabulary = vocabularies[feature_name]
            self._embeddings[feature_name] = tf.keras.Sequential([
                tf.keras.layers.experimental.preprocessing.StringLookup(
                    vocabulary=vocabulary, mask_token=None),
                tf.keras.layers.Embedding(len(vocabulary) + 1, self.embedding_dimension)
            ])
            
        for feature_name in int_features:
            vocabulary = vocabularies[feature_name]
            self._embeddings[feature_name] = tf.keras.Sequential([
                tf.keras.layers.experimental.preprocessing.IntegerLookup(
                    vocabulary=vocabulary, mask_value=None),
                tf.keras.layers.Embedding(len(vocabulary) + 1, self.embedding_dimension)
            ])

        if use_cross_layer:
            self._cross_layer = tfrs.layers.dcn.Cross(
                projection_dim=projection_dim,
                kernel_initializer="glorot_uniform")
        else:
            self._cross_layer = None
            
        self._deep_layers = [tf.keras.layers.Dense(layer_size, activation="relu")
                             for layer_size in deep_layer_sizes]
        
        self.task = tfrs.tasks.Ranking(
            loss=tf.keras.losses.MeanSquaredError(),
            metrics=[tf.keras.metrics.RootMeanSquaredError("RMSE")]
        )
        
        self.movie_model: tf.keras.layers.Layer = tf.keras.Sequential([
            tf.keras.layers.experimental.preprocessing.StringLookup(
                vocabulary=unique_movie_titles, mask_token=None),
            tf.keras.layers.Embedding(len(unique_movie_titles) + 1, self.embedding_dimension)
        ])
            
        self.rating_model: tf.keras.layers.Layer = tf.keras.Sequential([
            tf.keras.layers.Dense(256, activation="relu"),
            tf.keras.layers.Dense(128, activation="relu"),
            tf.keras.layers.Dense(1),
        ])
            
        self.rating_task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
            loss=tf.keras.losses.MeanSquaredError(),
            metrics=[tf.keras.metrics.RootMeanSquaredError()],
        )
        self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
            metrics=tfrs.metrics.FactorizedTopK(
                candidates=movies.batch(128).map(self.movie_model)
            )
        )
        self.rating_weight = rating_weight
        self.retrieval_weight = retrieval_weight
        
    def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
        embeddings = []
        for feature_name in self._all_features:
            embedding_fn = self._embeddings[feature_name]
            embeddings.append(embedding_fn(features[feature_name]))
        
        r_user_embeddings = tf.concat(embeddings, axis=1)
        
        if self._cross_layer is not None:
            r_user_embeddings = self._cross_layer(r_user_embeddings)
        
        for deep_layer in self._deep_layers:
            r_user_embeddings = deep_layer(r_user_embeddings)
        user_embeddings = self._embeddings["user_id"](features["user_id"])
        movie_embeddings = self.movie_model(features["movie_title"])
        
        return (
            user_embeddings,
            movie_embeddings,
            self.rating_model(
                tf.concat([r_user_embeddings, movie_embeddings], axis=1)
            ),
        )
    
    def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:
        ratings = features.pop("user_rating")
        
        user_embeddings, movie_embeddings, rating_predictions = self(features)
        
        rating_loss = self.rating_task(
            labels=ratings,
            predictions=rating_predictions,
        )
        retrieval_loss = self.retrieval_task(user_embeddings, movie_embeddings)
        
        return (self.rating_weight * rating_loss 
                + self.retrieval_weight * retrieval_loss)
</code></pre>
</details>

<details>
<summary>ScaNN and BruteForce Layers</summary>
<pre><code>
user_model = model._embeddings["user_id"]
brute_force = tfrs.layers.factorized_top_k.BruteForce(user_model)
brute_force.index(movies.batch(128).map(model.movie_model), movies)
<br>
scann = tfrs.layers.factorized_top_k.ScaNN(num_reordering_candidates=1000)
scann.index(movies.batch(128).map(model.movie_model), movies)
</code></pre>
</details>

<details>
<summary>Results</summary>
<pre><code>
# Get predictions for user 43.
_, titles = brute_force(np.array(["43"]), k=10)
print(f"Top recommendations: {titles[0]}")
</code>
Top recommendations: [b'Now and Then (1995)' b'Mr. Wrong (1996)'
 b'Affair to Remember, An (1957)' b'Cronos (1992)' b'Bed of Roses (1996)'
 b'Mrs. Winterbourne (1996)' b"Fathers' Day (1997)" b'Out to Sea (1997)'
 b'Simple Twist of Fate, A (1994)' b'Female Perversions (1996)']
<code>
_, titles = scann(user_model(np.array(["43"])), k=10)
print(f"Top recommendations: {titles[0]}")
</code>
Top recommendations: [b'Cronos (1992)' b'I.Q. (1994)'
 b'To Wong Foo, Thanks for Everything! Julie Newmar (1995)'
 b'Ref, The (1994)' b'Sneakers (1992)' b'One Fine Day (1996)'
 b'Manhattan Murder Mystery (1993)' b'Sabrina (1995)'
 b'Grand Day Out, A (1992)' b'Beautiful Girls (1996)']
</pre>
</details>

**Comment by maciejkula** [comment]
In general, the results aren't guaranteed to be the same - ScaNN is an approximation, so it may differ slightly.

Having said that, it does seem to differ a lot in your example. It's possible that your retrieval model is undertrained - have you considered temporarily disabling the ranking model (by setting its loss weight to 0) and seeing if that makes a difference?

(By the way, you can also pass the user model into the ScaNN layer constructor. Then calling both BruteForce and ScaNN will look the same.)

**Comment by saihtaungkham** [comment]
@maciejkula Thank you so much for your kind reply. I did an experiment with your suggestions and it came out that when  we disable the loss weight of ranking model, the output of ScaNN looks really great on the approximation of BruteForce layer. However, I am still having no clue what caused the retrieval model to be undertrained. Hope to get some suggestion from you guys. 

I also tried adding more Dense layers on the ranking model with the hope for the model to learn better. It is still not working.

<details>
<summary>FYI, I attached some train log and evaluation results.</summary>
<pre>
<code>
Epoch 42/50
10/10 [==============================] - 3s 318ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9548 - factorized_top_k/top_1_categorical_accuracy: 3.2500e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0373 - factorized_top_k/top_10_categorical_accuracy: 0.0777 - factorized_top_k/top_50_categorical_accuracy: 0.2885 - factorized_top_k/top_100_categorical_accuracy: 0.4448 - loss: 62256.2205 - regularization_loss: 0.0000e+00 - total_loss: 62256.2205
Epoch 43/50
10/10 [==============================] - 3s 316ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9325 - factorized_top_k/top_1_categorical_accuracy: 3.1250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0374 - factorized_top_k/top_10_categorical_accuracy: 0.0778 - factorized_top_k/top_50_categorical_accuracy: 0.2889 - factorized_top_k/top_100_categorical_accuracy: 0.4456 - loss: 62240.6005 - regularization_loss: 0.0000e+00 - total_loss: 62240.6005
Epoch 44/50
10/10 [==============================] - 3s 315ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9184 - factorized_top_k/top_1_categorical_accuracy: 3.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0376 - factorized_top_k/top_10_categorical_accuracy: 0.0781 - factorized_top_k/top_50_categorical_accuracy: 0.2893 - factorized_top_k/top_100_categorical_accuracy: 0.4464 - loss: 62225.5373 - regularization_loss: 0.0000e+00 - total_loss: 62225.5373
Epoch 45/50
10/10 [==============================] - 3s 319ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9378 - factorized_top_k/top_1_categorical_accuracy: 2.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0376 - factorized_top_k/top_10_categorical_accuracy: 0.0784 - factorized_top_k/top_50_categorical_accuracy: 0.2895 - factorized_top_k/top_100_categorical_accuracy: 0.4471 - loss: 62211.0238 - regularization_loss: 0.0000e+00 - total_loss: 62211.0238
Epoch 46/50
10/10 [==============================] - 3s 314ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9230 - factorized_top_k/top_1_categorical_accuracy: 2.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0377 - factorized_top_k/top_10_categorical_accuracy: 0.0786 - factorized_top_k/top_50_categorical_accuracy: 0.2902 - factorized_top_k/top_100_categorical_accuracy: 0.4479 - loss: 62196.9148 - regularization_loss: 0.0000e+00 - total_loss: 62196.9148
Epoch 47/50
10/10 [==============================] - 3s 322ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9212 - factorized_top_k/top_1_categorical_accuracy: 2.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0379 - factorized_top_k/top_10_categorical_accuracy: 0.0786 - factorized_top_k/top_50_categorical_accuracy: 0.2907 - factorized_top_k/top_100_categorical_accuracy: 0.4484 - loss: 62183.2756 - regularization_loss: 0.0000e+00 - total_loss: 62183.2756
Epoch 48/50
10/10 [==============================] - 3s 312ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9276 - factorized_top_k/top_1_categorical_accuracy: 2.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0379 - factorized_top_k/top_10_categorical_accuracy: 0.0788 - factorized_top_k/top_50_categorical_accuracy: 0.2914 - factorized_top_k/top_100_categorical_accuracy: 0.4492 - loss: 62170.0614 - regularization_loss: 0.0000e+00 - total_loss: 62170.0614
Epoch 49/50
10/10 [==============================] - 3s 319ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9181 - factorized_top_k/top_1_categorical_accuracy: 2.7500e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0381 - factorized_top_k/top_10_categorical_accuracy: 0.0790 - factorized_top_k/top_50_categorical_accuracy: 0.2918 - factorized_top_k/top_100_categorical_accuracy: 0.4501 - loss: 62157.2184 - regularization_loss: 0.0000e+00 - total_loss: 62157.2184
Epoch 50/50
10/10 [==============================] - 3s 324ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9187 - factorized_top_k/top_1_categorical_accuracy: 2.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0382 - factorized_top_k/top_10_categorical_accuracy: 0.0791 - factorized_top_k/top_50_categorical_accuracy: 0.2924 - factorized_top_k/top_100_categorical_accuracy: 0.4508 - loss: 62144.7670 - regularization_loss: 0.0000e+00 - total_loss: 62144.7670
</code>
</pre>
</details>

<details>
<summary>Results on evaluation dataset.</summary>
<pre>
<code>
5/5 [==============================] - 6s 216ms/step - RMSE: 0.0000e+00 - root_mean_squared_error: 0.9308 - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0011 - factorized_top_k/top_10_categorical_accuracy: 0.0038 - factorized_top_k/top_50_categorical_accuracy: 0.0692 - factorized_top_k/top_100_categorical_accuracy: 0.1730 - loss: 32792.2337 - regularization_loss: 0.0000e+00 - total_loss: 32792.2337
Retrieval top-100 accuracy: 0.173.
Ranking RMSE: 0.931.
</code>
</pre>
</details>

**Comment by saihtaungkham** [comment]
@maciejkula I have figured out the problem. After tuning the ScaNN with the following configurations, it works. Thank you for the suggestion and support!

```
scann = tfrs.layers.factorized_top_k.ScaNN(
    query_model=user_model, 
    num_leaves_to_search=70,
    num_reordering_candidates=400
)
```
